File: 06-social-media-scripts/docs/research/ig-tools-bench/bench_profile.mjs
/** Workload real: traer los últimos N reels de una cuenta (North Star Account). */
import { igApi } from 'insta-fetcher';
import fs from 'node:fs';

const N = 12;
const ACCOUNTS = ['nasa', 'openai'];
const SESSIONID = (process.env.IG_SESSIONID ?? '').trim();
const ig = new igApi(SESSIONID ? `sessionid=${SESSIONID};` : undefined);
const sec = (t) => Number(((performance.now() - t) / 1000).toFixed(2));

const out = [];
for (const acc of ACCOUNTS) {
  const row = { account: acc };
  try {
    let t = performance.now();
    const reel = await ig.fetchUserReel(acc, null, N);
    row.t_total_s = sec(t);
    // shape real (2026): xdt_api__v1__clips__user__connection_v2.edges[].node.media
    const edges = reel?.xdt_api__v1__clips__user__connection_v2?.edges
      ?? reel?.items ?? reel?.data?.items ?? [];
    const media = edges.map(e => e?.node?.media ?? e?.media ?? e).filter(Boolean);
    row.reels = media.length;
    row.s_per_reel = Number((row.t_total_s / Math.max(media.length, 1)).toFixed(2));
    row.status = media.length ? 'ok' : 'empty';
    const m = media[0];
    row.sample = m ? {
      shortcode: m.code, likes: m.like_count, comments: m.comment_count,
      play_count: m.play_count ?? m.ig_play_count, duration: m.video_duration,
      caption_len: (m.caption?.text ?? '').length,
      music: m.clips_metadata?.music_info?.music_asset_info?.title ?? m.clips_metadata?.original_sound_info?.original_audio_title,
      video_url: !!m.video_versions?.[0]?.url,
    } : null;
  } catch (e) {
    row.status = 'error';
    row.error = `${e?.name}: ${String(e?.message).slice(0, 200)}`;
  }
  console.log(JSON.stringify(row));
  out.push(row);
}

fs.writeFileSync('results/profile_instafetcher.json',
  JSON.stringify({ tool: 'insta-fetcher', workload: `últimos ${N} reels por cuenta`, results: out }, null, 1));

File: 06-social-media-scripts/docs/research/ig-tools-bench/bench_profile.py
"""Workload real: traer los últimos N reels de una cuenta (North Star Account)."""
import os, time, json
import instaloader

N = 12
ACCOUNTS = ["nasa", "openai"]
SESSIONID = os.environ.get("IG_SESSIONID", "").strip()

L = instaloader.Instaloader(quiet=True, save_metadata=False, download_comments=False)
if SESSIONID:
    L.context._session.cookies.set("sessionid", SESSIONID, domain=".instagram.com")
    try:
        L.context.username = L.test_login()
    except Exception as e:
        print("[warn] test_login:", e)

out = []
for acc in ACCOUNTS:
    t0 = time.perf_counter()
    try:
        prof = instaloader.Profile.from_username(L.context, acc)
        t_prof = time.perf_counter() - t0
        t1 = time.perf_counter()
        reels, n_req = [], 0
        for p in prof.get_posts():
            if p.is_video:
                reels.append({"shortcode": p.shortcode, "likes": p.likes, "comments": p.comments,
                              "caption_len": len(p.caption or ""), "date": p.date_utc.isoformat(),
                              "video_url": bool(p.video_url)})
            n_req += 1
            if len(reels) >= N or n_req >= 30:
                break
        t_posts = time.perf_counter() - t1
        row = {"account": acc, "status": "ok", "t_profile_s": round(t_prof, 2),
               "t_posts_s": round(t_posts, 2), "t_total_s": round(t_prof + t_posts, 2),
               "reels": len(reels), "posts_scanned": n_req,
               "s_per_reel": round((t_prof + t_posts) / max(len(reels), 1), 2)}
    except Exception as e:
        row = {"account": acc, "status": "error", "error": f"{type(e).__name__}: {str(e)[:200]}",
               "t_total_s": round(time.perf_counter() - t0, 2)}
    print(json.dumps(row, ensure_ascii=False))
    out.append(row)

json.dump({"tool": "instaloader", "workload": f"últimos {N} reels por cuenta", "results": out},
          open("results/profile_instaloader.json", "w"), indent=1, ensure_ascii=False)
